velero-plugin-for-microsoft-azure/object_store.go

```diff
@@ -235,16 +248,82 @@ func (o *ObjectStore) Init(config map[string]string) error {
 		blobService: &blobClient,
 	}
 
+	o.blockSize = getBlockSize(o.log, config)
+
 	return nil
 }
 
+func getBlockSize(log logrus.FieldLogger, config map[string]string) int {
+	val, ok := config[blockSizeConfigKey]
+	if !ok {
+		// no alternate block size specified in config, so return with the default
+		return defaultBlockSize
+	}
+
+	blockSize, err := strconv.Atoi(val)
+	if err != nil {
+		log.WithError(err).Warnf("Error parsing config.blockSizeInBytes value %v, using default block size of %d", val, defaultBlockSize)
+		return defaultBlockSize
+	}
+
+	if blockSize <= 0 || blockSize > defaultBlockSize {
+		log.WithError(err).Warnf("Value provided for config.blockSizeInBytes (%d) is outside the allowed range of 1 to %d, using default block size of %d", blockSize, defaultBlockSize, defaultBlockSize)
+		return defaultBlockSize
+	}
+
+	return blockSize
+}
+
 func (o *ObjectStore) PutObject(bucket, key string, body io.Reader) error {
 	blob, err := o.blobGetter.getBlob(bucket, key)
 	if err != nil {
 		return err
 	}
 
-	return errors.WithStack(blob.CreateBlockBlobFromReader(body, nil))
+	// Azure requires a blob/object to be chunked if it's larger than 256MB. Since we
+	// don't know ahead of time if the body is over this limit or not, and it would
+	// require reading the entire object into memory to determine the size, we use the
+	// chunking approach for all objects.
+
+	var (
+		block    = make([]byte, o.blockSize)
+		blockIDs []storage.Block
+	)
+
+	for {
+		n, err := body.Read(block)
+		if n > 0 {
+			// blockID needs to be the same length for all blocks, so use a fixed width.
+			// ref. https://docs.microsoft.com/en-us/rest/api/storageservices/put-block#uri-parameters
+			blockID := fmt.Sprintf("%08d", len(blockIDs))
+
+			o.log.Debugf("Putting block (id=%s) of length %d", blockID, n)
+			if putErr := blob.PutBlock(blockID, block[0:n], nil); putErr != nil {
+				return errors.Wrapf(putErr, "error putting block %s", blockID)
+			}
+
+			blockIDs = append(blockIDs, storage.Block{
+				ID:     blockID,
+				Status: storage.BlockStatusLatest,
+			})
+		}
+
+		// got an io.EOF: we're done reading chunks from the body
+		if err == io.EOF {
+			break
+		}
+		// any other error: bubble it up
+		if err != nil {
+			return errors.Wrap(err, "error reading block from body")
+		}
+	}
+
+	o.log.Debugf("Putting block list %v", blockIDs)
```

Comment:
Out of curiosity.. if  our app crashes before doing this. I assume there is no way to recover from it and the blob store will hold on to our blocks that we put for sometime discard them if it doesn't receive a `PutBlockList` call.

Reply:
the other question would be: what's the max allowed time between all the `PutBlocks` and the `PutBlockList`?
Should the plugin be concerned with that?

Reply:
This is the best reference I've found for how it works: https://docs.microsoft.com/en-us/rest/api/storageservices/put-block#remarks

Specifically: "The block or blocks that you uploaded are not committed until you call Put Block List on the new blob. A blob created this way is maintained on the server for a week; if you have not added more blocks or committed blocks to the blob within that time period, then the blob is garbage collected."